gdal/frmts/gtiff/libtiff/tif_webp.c

```diff
@@ -0,0 +1,520 @@
+	(void) s;
+  WebPState *sp = EncoderState(tif);
+  assert(sp != NULL);
+  assert(sp->state == LSTATE_INIT_ENCODE);
+  int64_t stride = (int64_t)sp->sPicture.width * sp->nSamples;
+
+  if ((stride * sp->sPicture.height) != cc)
+  {
+    TIFFErrorExt(tif->tif_clientdata, module, "Cannot encode WebP with buffers this size");
+    return 0;
+  }
+
+#if WEBP_ENCODER_ABI_VERSION >= 0x0100
+  if (sp->nSamples == 4)
+  {
+      if (!WebPPictureImportRGBA(&sp->sPicture, bp, (int)stride))
+      {
+          TIFFErrorExt(tif->tif_clientdata, module,
+                    "WebPPictureImportRGBA() failed" );
+          eErr = sp->sPicture.error_code;
+      }
+  }
+  else
+#endif
+  if ((eErr == VP8_ENC_OK) &&
+      !WebPPictureImportRGB(&sp->sPicture, bp, (int)stride))
+  {
+      TIFFErrorExt(tif->tif_clientdata, module,
+                    "WebPPictureImportRGB() failed");
+      eErr = sp->sPicture.error_code;
+  }
+
+  if ((eErr == VP8_ENC_OK) && !WebPEncode(&sp->sEncoderConfig, &sp->sPicture))
+  {
+    // copied from webpdataset.cpp
+#if WEBP_ENCODER_ABI_VERSION >= 0x0100
+    const char* pszErrorMsg = NULL;
+    switch(sp->sPicture.error_code)
+    {
+        case VP8_ENC_ERROR_OUT_OF_MEMORY:
+            pszErrorMsg = "Out of memory"; break;
+        case VP8_ENC_ERROR_BITSTREAM_OUT_OF_MEMORY:
+            pszErrorMsg = "Out of memory while flushing bits"; break;
+        case VP8_ENC_ERROR_NULL_PARAMETER:
+            pszErrorMsg = "A pointer parameter is NULL"; break;
+        case VP8_ENC_ERROR_INVALID_CONFIGURATION:
+            pszErrorMsg = "Configuration is invalid"; break;
+        case VP8_ENC_ERROR_BAD_DIMENSION:
+            pszErrorMsg = "Picture has invalid width/height"; break;
+        case VP8_ENC_ERROR_PARTITION0_OVERFLOW:
+            pszErrorMsg = "Partition is bigger than 512k. Try using less "
+                "SEGMENTS, or increase PARTITION_LIMIT value";
+            break;
+        case VP8_ENC_ERROR_PARTITION_OVERFLOW:
+            pszErrorMsg = "Partition is bigger than 16M";
+            break;
+        case VP8_ENC_ERROR_BAD_WRITE:
+            pszErrorMsg = "Error while flushing bytes"; break;
+        case VP8_ENC_ERROR_FILE_TOO_BIG:
+            pszErrorMsg = "File is bigger than 4G"; break;
+        case VP8_ENC_ERROR_USER_ABORT:
+            pszErrorMsg = "User interrupted";
+            break;
+        default:
+            TIFFErrorExt(tif->tif_clientdata, module,
+                    "WebPEncode returned an unknown error code: %d",
+                    sp->sPicture.error_code);
+            pszErrorMsg = "Unknown WebP error type.";
+            break;
+    }
+    TIFFErrorExt(tif->tif_clientdata, module,
+             "WebPEncode() failed : %s", pszErrorMsg);
+#else
+    TIFFErrorExt(tif->tif_clientdata, module,
+             "Error in WebPEncode()");
+#endif
+    return 0;
+  }
+
+  bp += cc;
+  cc = 0;
+
+	return 1;
+}
+
+static int
+TWebPDecode(TIFF* tif, uint8* op, tmsize_t occ, uint16 s)
```

Comment:
in case of very long strips/tiles, in the CHUNKY_STRIP_READ_SUPPORT mode used for GDAL compilation, this function should be called several times with parts of the strips/tile. Tests should be made with super big strips/tiles to detect that and error out if we don't want/ can't support it

Reply:
We should check somewhere that PlanarConfig == PLANARCONFIG_CONTIG

Reply:
libwebp supports this on the decode as it is possible to decode in chunks using `WebPIDecoder` but I didn't see a way to encode in chunks which would have been nice. I guess you could create the file one one system and then want to read it on another using less memory, but I will error out here for now.

Reply:
+1 added a check for PLANARCONFIG_CONTIG in TWebPFixupTags following the pattern in tif_jpeg

Reply:
WebPIDecoder seems to keep the buffer for the entire image which makes chunky read support redundant, reverted to checking the header of each call to decode to see if we are at the start and raising an error if not.